wp-content/themes/core/components/accordion/Controller.php

```diff
@@ -0,0 +1,142 @@
+ *
+ * This component has these features out of box:
+ *  - Full accessibility baked in.
+ *  - Lightweight CSS based height animations
+ *  - One item open at a time, with scrolling to keep items on screen. Easily switchable in the js.
+ *
+ * rows take an array if items. Each row item should look like this.
+ * [
+ *     'header_text' => 'The Row Title',
+ *     'header_id' => 'uid-123',
+ *     'content_id' => 'uid-123',
+ *     'content => 'Amazing accordion content',
+ * ]
+ */
+class Controller extends Abstract_Controller {
+
+	/**
+	 * @var string[]s
+	 */
+	public $rows;
+
+	/**
+	 * @var string[]
+	 */
+	public $container_classes;
+
+	/**
+	 * @var string[]
+	 */
+	public $container_attrs;
+
+	/**
+	 * @var string[]
+	 */
+	public $row_classes;
+
+	/**
+	 * @var string
+	 */
+	public $row_header_tag;
+
+	/**
+	 * @var string[]
+	 */
+	public $row_header_classes;
+
+	/**
+	 * @var string[]
+	 */
+	public $row_header_container_classes;
+
+	/**
+	 * @var string[]
+	 */
+	public $row_content_classes;
+
+	/**
+	 * @var string[]
+	 */
+	public $row_content_container_classes;
+
+	/**
+	 * @var string
+	 */
+	public $row_header_name;
+
+	/**
+	 * @var string
+	 */
+	public $row_content_name;
+
+	public function __construct( array $args ) {
+		$args                                = wp_parse_args( $args, $this->defaults() );
+		$this->rows                          = (array) $args[ 'rows' ];
+		$this->container_classes             = (array) $args[ 'container_classes' ];
+		$this->container_attrs               = (array) $args[ 'container_attrs' ];
+		$this->row_header_tag                = $args[ 'row_header_tag' ];
+		$this->row_classes                   = (array) $args[ 'row_classes' ];
+		$this->row_header_classes            = (array) $args[ 'row_header_classes' ];
+		$this->row_header_container_classes  = (array) $args[ 'row_header_container_classes' ];
+		$this->row_content_classes           = (array) $args[ 'row_content_classes' ];
+		$this->row_content_container_classes = (array) $args[ 'row_content_container_classes' ];
+		$this->row_header_name               = $args[ 'row_header_name' ];
+		$this->row_content_name              = $args[ 'row_content_name' ];
+	}
+
+	protected function defaults(): array {
+		return [
+			'rows'                          => [],
+			'container_classes'             => [ 'c-accordion' ],
+			'container_attrs'               => [ 'data-js' => 'c-accordion' ],
+			'row_header_tag'                => 'h3',
+			'row_classes'                   => [ 'c-accordion__row' ],
+			'row_header_classes'            => [ 'c-accordion__header', 'h5' ],
+			'row_header_container_classes'  => [ 'c-accordion__header-container' ],
+			'row_content_classes'           => [ 'c-accordion__content' ],
+			'row_content_container_classes' => [ 'c-accordion__content-container', 't-sink', 's-sink' ],
+			'row_header_name'               => 'title',
+			'row_content_name'              => 'row_content',
+		];
+	}
```

Comment:
These defaults can be setup as part of the constructor above (scope out any of the other done components).

Reply:
Fair enough, it is cleaner more clear. I think the main reason I've liked the other is the reduction in code in the controller. I could go either way. @nickpelton @dpellenwood @jbrinley?

Reply:
I'm also finding this true as I've worked on the Tabs component. I agree with you, Mike. That being said, we need to keep the ability to both allow defaults and required classes/attributes for each component. In this case, the majority of the values here should be required, not just defaults as the component's JS requires the attrs to function properly and the component's styles rely on these classes.

Reply:
For context, this is the tabs component with the current arch (no defaults method) showing the required and default properties all inlined. I think this is really, really hard to read:
```
	public function __construct( array $args = [] ) {
		$this->tabs                = (array) $args['tabs'] ?? [];
		$this->container_classes   = array_merge( ['c-tabs'], (array) $args['container_classes'] ?? [] );
		$this->container_attrs     = array_merge( [ 'data-js' => 'c-tabs' ], (array) $args['container_attrs'] ?? [] );
		$this->tab_list_classes    = array_merge( ['c-tab__list'], (array) $args['tab_list_classes'] ?? [] );
		$this->tab_list_attrs      = (array) $args['tab_list_attrs'] ?? [];
		$this->tab_button_classes  = array_merge( ['c-tab__button'], (array) $args['tab_button_classes'] ?? [] );
		$this->tab_button_attrs    = (array) $args['tab_button_attrs'] ?? [];
		$this->tab_content_classes = array_merge( ['c-tab__content'], (array) $args['tab_button_classes'] ?? ['s-sink', 't-sink'] );
		$this->tab_content_attrs   = array_merge( [ 'role' => 'tabpanel', 'tabindex' => 0 ], (array) $args['tab_content_attrs'] ?? [] );
		$this->tab_list_buttons    = $this->get_tablist_buttons();
	}
```

Reply:
I also like this style better. The heavy use of the `??` was mostly for speed before we go back and refactor common problems back into the `Abstract_Controller`. After doing it enough times, I think this approach makes sense as long as we can still have defaults, override defaults, and some required arguments.

Reply:
Proposal: We setup the defaults this way and inject the required properties within the markup util methods (below)?

Reply:
Plus one to this style. What about having two conventions?

1. A `defaults()` method that returns an array of defaults that can be overwritten with the args.
2. A `merge()` method that returns an array of arrays that will be merged with the args.

Reply:
Yep. This is better, Jonathan. I was just getting to this as I was adjusting the tabs component. My idea of merging inside the markup utils wasn't ideal either.